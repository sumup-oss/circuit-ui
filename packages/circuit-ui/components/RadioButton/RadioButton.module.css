.label {
  position: relative;
  display: inline-block;
  padding-left: 26px;
  color: var(--cui-fg-normal);
  cursor: pointer;
}

.label::before {
  position: absolute;
  top: calc(var(--cui-body-m-line-height) / 2);
  left: 0;
  box-sizing: border-box;
  display: block;
  width: 18px;
  height: 18px;
  content: "";
  background-color: var(--cui-bg-normal);
  border: 1px solid var(--cui-border-normal);
  border-radius: 100%;
  box-shadow: none;
  transition:
    border var(--cui-transitions-default),
    background-color var(--cui-transitions-default);
  transform: translateY(-50%);
}

.label::after {
  position: absolute;
  top: calc(var(--cui-body-m-line-height) / 2);
  left: var(--cui-spacings-bit);
  box-sizing: border-box;
  display: block;
  width: 10px;
  height: 10px;
  content: "";
  background-color: var(--cui-fg-accent);
  border-radius: 100%;
  opacity: 0;
  transition:
    transform var(--cui-transitions-default),
    opacity var(--cui-transitions-default);
  transform: translateY(-50%) scale(0, 0);
}

.base:hover + label::before {
  border-color: var(--cui-border-normal-hovered);
}

.base:focus + label::before {
  border-color: var(--cui-border-accent);
  outline: 0;
  box-shadow:
    0 0 0 2px var(--cui-bg-normal),
    0 0 0 4px var(--cui-border-focus);
}

.base:focus:not(:focus-visible) + label::before {
  border-color: var(--cui-border-normal);
  box-shadow: none;
}

.base:checked:focus:not(:focus-visible) + label::before {
  border-color: var(--cui-border-accent);
}

.base:checked + label::before {
  border-color: var(--cui-border-accent);
}

.base:checked + label::after {
  opacity: 1;
  transform: translateY(-50%) scale(1, 1);
}

.base:disabled + label,
.base[disabled] + label {
  color: var(--cui-fg-normal-disabled);
  pointer-events: none;
}

.base:disabled + label::before,
.base[disabled] + label::before {
  background-color: var(--cui-bg-normal-disabled);
  border-color: var(--cui-border-normal-disabled);
}

.base:disabled + label::after,
.base[disabled] + label::after {
  background-color: var(--cui-fg-on-strong-disabled);
}

.base:disabled:checked + label::before,
.base[disabled]:checked + label::before {
  border-color: var(--cui-border-accent-disabled);
}

.base:disabled:checked + label::after,
.base[disabled]:checked + label::after {
  background-color: var(--cui-fg-accent-disabled);
}

/* Invalid */

[aria-invalid="true"] .base:not(:focus) + .label::before {
  background-color: var(--cui-bg-danger);
  border-color: var(--cui-border-danger);
}

[aria-invalid="true"] .base:not(:focus) + .label::after {
  background-color: var(--cui-fg-danger);
}

[aria-invalid="true"] .base:hover + .label::before,
[aria-invalid="true"] .base:focus + .label::before {
  border-color: var(--cui-border-danger-hovered);
}

[aria-invalid="true"] .base:checked + .label::before {
  border-color: var(--cui-border-danger);
}
