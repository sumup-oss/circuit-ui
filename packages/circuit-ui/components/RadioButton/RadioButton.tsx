/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/** @jsx jsx */
import React, { Fragment, HTMLProps, Ref } from 'react';
import { css, jsx } from '@emotion/core';
import { Dispatch as TrackingProps } from '@sumup/collector';

import styled, { StyleProps } from '../../styles/styled';
import {
  disableVisually,
  hideVisually,
  focusOutline,
} from '../../styles/style-mixins';
import { uniqueId } from '../../util/id';
import useClickHandler from '../../hooks/use-click-handler';

export interface RadioButtonProps extends HTMLProps<HTMLInputElement> {
  /**
   * Triggers error styles on the component.
   */
  invalid?: boolean;
  /**
   * The ref to the HTML Dom element
   */
  ref?: Ref<HTMLInputElement>;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  tracking?: TrackingProps;
}

type LabelElProps = Pick<RadioButtonProps, 'invalid' | 'disabled'>;

const labelBaseStyles = ({ theme }: StyleProps) => css`
  label: radio-button__label;
  color: ${theme.colors.bodyColor};
  padding-left: 26px;
  position: relative;
  cursor: pointer;

  &::before {
    box-sizing: border-box;
    height: 18px;
    width: 18px;
    box-shadow: 0;
    background-color: ${theme.colors.white};
    border: 1px solid ${theme.colors.n500};
    border-radius: 100%;
    content: '';
    display: block;
    position: absolute;
    top: 50%;
    left: 0;
    transform: translateY(-50%);
    transition: border ${theme.transitions.default};
  }

  &::after {
    box-sizing: border-box;
    height: 10px;
    width: 10px;
    background-color: ${theme.colors.p500};
    border-radius: 100%;
    content: '';
    display: block;
    position: absolute;
    top: 50%;
    left: ${theme.spacings.bit};
    transform: translateY(-50%) scale(0, 0);
    opacity: 0;
    transition: transform ${theme.transitions.default},
      opacity ${theme.transitions.default};
  }
`;

const labelInvalidStyles = ({ theme, invalid }: StyleProps & LabelElProps) =>
  invalid &&
  css`
    label: radio-button--error;
    &:not(:focus)::before {
      border-color: ${theme.colors.danger};
      background-color: ${theme.colors.r100};
    }

    &:not(:focus)::after {
      background-color: ${theme.colors.danger};
    }
  `;

const labelDisabledStyles = ({ theme, disabled }: StyleProps & LabelElProps) =>
  disabled &&
  css`
    label: radio-button--disabled;
    ${disableVisually()};

    &::before {
      ${disableVisually()};
      border-color: ${theme.colors.n700};
      background-color: ${theme.colors.n200};
    }

    &::after {
      ${disableVisually()};
      background-color: ${theme.colors.n700};
    }
  `;

const RadioButtonLabel = styled('label')<LabelElProps>(
  labelBaseStyles,
  labelDisabledStyles,
  labelInvalidStyles,
);

type InputElProps = Omit<RadioButtonProps, 'tracking'>;

const inputBaseStyles = ({ theme }: StyleProps) => css`
  label: radio-button__input;
  ${hideVisually()};

  &:hover + label::before {
    border-color: ${theme.colors.n700};
  }

  &:focus + label::before {
    ${focusOutline({ theme })};
    border-color: ${theme.colors.p500};
  }

  &:checked + label {
    &::before {
      border-color: ${theme.colors.p500};
    }

    &::after {
      transform: translateY(-50%) scale(1, 1);
      opacity: 1;
    }
  }
`;

const inputInvalidStyles = ({ theme, invalid }: StyleProps & InputElProps) =>
  invalid &&
  css`
    label: radio-button__input--invalid;

    &:hover + label::before,
    &:focus + label::before {
      border-color: ${theme.colors.r700};
    }

    &:checked + label::before {
      border-color: ${theme.colors.danger};
    }
  `;

const RadioButtonInput = styled('input')<InputElProps>(
  inputBaseStyles,
  inputInvalidStyles,
);

/**
 * RadioButton component for forms.
 */
export const RadioButton = React.forwardRef(
  (
    {
      onChange,
      children,
      label,
      id: customId,
      name,
      value,
      checked,
      invalid,
      disabled,
      tracking,
      className,
      style,
      ...props
    }: RadioButtonProps,
    ref: RadioButtonProps['ref'],
  ) => {
    const id = customId || uniqueId('radio-button_');
    const handleChange = useClickHandler(onChange, tracking, 'radio-button');

    return (
      <Fragment>
        <RadioButtonInput
          {...props}
          type="radio"
          name={name}
          id={id}
          value={value}
          invalid={invalid}
          disabled={disabled}
          checked={checked}
          onClick={handleChange}
          onChange={() => {
            /**
             * Noop to silence React warning:
             * https://github.com/facebook/react/issues/3070#issuecomment-73311114
             * Change is handled by onClick which has better browser support:
             * https://stackoverflow.com/a/5575369/4620154
             */
          }}
          ref={ref}
        />
        <RadioButtonLabel
          htmlFor={id}
          disabled={disabled}
          invalid={invalid}
          className={className}
          style={style}
        >
          {children || label}
        </RadioButtonLabel>
      </Fragment>
    );
  },
);

RadioButton.displayName = 'RadioButton';
