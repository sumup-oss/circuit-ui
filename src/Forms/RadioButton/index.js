import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './index.scss';
import withStyles from '../../../util/withStyles';

const RadioButton = ({
  children,
  id,
  name,
  value,
  disabled,
  onChange,
  checked,
  className
}) => (
  <div
    className={classNames('radio-button', className, {
      'radio-button--disabled': disabled
    })}
  >
    <input
      type="radio"
      name={name}
      id={id || name}
      disabled={disabled}
      onChange={onChange}
      value={value}
      checked={checked}
    />
    {children}
  </div>
);

RadioButton.propTypes = {
  name: PropTypes.string,
  id: PropTypes.string,
  value: PropTypes.string,
  disabled: PropTypes.bool,
  checked: PropTypes.bool,
  onChange: PropTypes.func,
  className: PropTypes.string
};

export default withStyles(styles)(RadioButton);
