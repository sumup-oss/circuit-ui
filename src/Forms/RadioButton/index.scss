@import '../../shared';

.radio-button {
  label {
    cursor: pointer;
    display: inline-block;
    font-size: $font-size-m;
    font-weight: 300;
    margin-right: 15px;
    padding-left: 25px;
    padding-right: 15px;
    position: relative;

    &::before {
      background-color: $color-white;
      border: 1px solid $color-medium;
      content: '';
      display: inline-block;
      height: 16px;
      left: 0;
      margin-right: 10px;
      position: absolute;
      top: 0;
      width: 16px;
    }
  }

  input[type='radio'] { /* stylelint-disable-line selector-no-qualifying-type */
    border: 0;
    clip: auto;
    height: auto;
    margin: 0;
    overflow: visible;
    padding: 0;
    position: absolute;
    visibility: hidden;
    width: auto;
  }

  input[type='radio'] + label::before { /* stylelint-disable-line selector-no-qualifying-type */
    border-radius: 9px;
  }

  input:checked + label::after {
    display: inline-block;
    height: 8px;
    position: absolute;
    width: 8px;
  }

  input[type='radio']:checked + label::after { /* stylelint-disable-line selector-no-qualifying-type */
    background: $color-brand;
    border-radius: 4px;
    content: '';
    left: 4px;
    top: 4px;
  }

  .radio-button--disabled {
    opacity: .4;

    label {
      cursor: default;
    }
  }
}
