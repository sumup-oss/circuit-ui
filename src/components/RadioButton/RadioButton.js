/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import {
  disableVisually,
  hideVisually,
  focusOutline
} from '../../styles/style-helpers';
import { childrenPropType } from '../../util/shared-prop-types';
import { uniqueId } from '../../util/id';

const labelBaseStyles = ({ theme }) => css`
  label: radio-button__label;
  color: ${theme.colors.n700};
  padding-left: ${theme.spacings.giga};
  position: relative;
  cursor: pointer;

  &::before {
    box-sizing: border-box;
    height: ${theme.spacings.mega};
    width: ${theme.spacings.mega};
    box-shadow: inset 0 1px 2px 0 rgba(102, 113, 123, 0.12);
    background-color: ${theme.colors.white};
    border: 1px solid ${theme.colors.n500};
    border-radius: 100%;
    content: '';
    display: block;
    position: absolute;
    top: 50%;
    left: 0;
    transform: translateY(-50%);
    transition: border 0.05s ease-in;
  }

  &::after {
    box-sizing: border-box;
    height: ${theme.spacings.byte};
    width: ${theme.spacings.byte};
    background-color: ${theme.colors.p500};
    border-radius: 100%;
    content: '';
    display: block;
    position: absolute;
    top: 50%;
    left: ${theme.spacings.bit};
    transform: translateY(-50%) scale(0, 0);
    opacity: 0;
    transition: transform 0.05s ease-in, opacity 0.05s ease-in;
  }
`;

const labelInvalidStyles = ({ theme, invalid }) =>
  invalid &&
  css`
    label: radio-button--error;
    &:not(:focus)::before {
      border-color: ${theme.colors.r500};
      background-color: ${theme.colors.r100};
    }

    &:not(:focus)::after {
      background-color: ${theme.colors.r500};
    }
  `;

const labelDisabledStyles = ({ theme, disabled }) =>
  disabled &&
  css`
    label: radio-button--disabled;
    ${disableVisually()};

    &::before {
      ${disableVisually()};
      border-color: ${theme.colors.n500};
      background-color: ${theme.colors.n100};
    }

    &::after {
      ${disableVisually()};
      background-color: ${theme.colors.n500};
    }
  `;

const inputStyles = ({ theme }) => css`
  label: radio-button__input;

  &:focus + label::before {
    ${focusOutline({ theme })}
  }

  &:checked + label {
    &::before {
      border-color: ${theme.colors.p500};
    }

    &::after {
      transform: translateY(-50%) scale(1, 1);
      opacity: 1;
    }
  }
`;

const RadioButtonInput = styled('input')(hideVisually, inputStyles);

const RadioButtonLabel = styled('label')(
  labelBaseStyles,
  labelDisabledStyles,
  labelInvalidStyles
);

/**
 * RadioButton component for forms.
 */
const RadioButton = ({
  onChange,
  children,
  id,
  name,
  value,
  checked,
  ...props
}) => {
  const inputId = id || uniqueId('radio-button_');
  return (
    <>
      <RadioButtonInput
        {...props}
        type="radio"
        name={name}
        id={inputId}
        value={value}
        checked={checked}
        onClick={onChange}
      />
      <RadioButtonLabel {...props} htmlFor={inputId}>
        {children}
      </RadioButtonLabel>
    </>
  );
};

RadioButton.propTypes = {
  /**
   * Callback used when the user toggles the radio button.
   */
  onChange: PropTypes.func,
  /**
   * Value string for input.
   */
  value: PropTypes.string.isRequired,
  /**
   * Child nodes to be rendered as the label.
   */
  children: childrenPropType.isRequired,
  /**
   * The name of the radio group that the radio button belongs to.
   */
  name: PropTypes.string.isRequired,
  /**
   * A unique ID used to link the input and label.
   */
  id: PropTypes.string,
  /**
   * Triggers checked styles on the component. This is also forwarded as
   * attribute to the <input> element.
   */
  checked: PropTypes.bool,
  /**
   * Triggers error styles on the component.
   */
  invalid: PropTypes.bool,
  /**
   * Triggers disabled styles on the component. This is also forwarded as
   * attribute to the <input> element.
   */
  disabled: PropTypes.bool
};

RadioButton.defaultProps = {
  id: null,
  checked: false,
  invalid: false,
  disabled: false,
  children: null
};

/**
 * @component
 */
export default RadioButton;
