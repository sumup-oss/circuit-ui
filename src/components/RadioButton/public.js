export { default as RadioButton } from '.';
