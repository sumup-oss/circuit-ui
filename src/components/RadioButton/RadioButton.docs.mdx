import { Status, Props, Story } from '../../../.storybook/components';
import RadioButton from './RadioButton';
import RadioButtonGroup from '../RadioButtonGroup';

# RadioButton

<Status.Stable />

A radio button is a quick way to extract a single coded answer from our users by presenting multiple
possibilities and allowing only one option to be chosen.

<Story id="forms-radiobutton--base" />

<Props of={RadioButton} />

## When to use it

Use a radio button when the user has to choose only one option based on a pre-defined list of mutually
exclusive options.

## Usage guidelines

- **Do** always associate the radio button with a label that clearly gives context for the user to make a decision
- **Do** capitalize each item of the list
- **Do** list the items in a rational order
- **Do not** use more than 4 options on a list of radio buttons. If it exceeds that, use a selector.
- **Do not** punctuate the end of every list item
- **Do not** mark required fields with asterisks

## Component variations

### Disabled Radio Button

Use a disabled radio button only if you really need to communicate to the user that an option that existed
before is not available for choosing now. Keep in mind that you need to communicate why. One recommendation
would be to pair a disabled input with a tooltip to give the appropriate context for our user.

<Story id="forms-radiobutton--disabled" />

### Invalid

<Story id="forms-radiobutton--invalid" />

## RadioButtonGroup

<Story id="forms-radiobutton-radiobuttongroup--base" />

<Props of={RadioButtonGroup} />
