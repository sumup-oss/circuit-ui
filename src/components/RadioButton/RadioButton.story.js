/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';

import RadioButton from './RadioButton';
import docs from './RadioButton.docs.mdx';

export default {
  title: 'Forms|RadioButton',
  component: RadioButton,
  parameters: {
    docs: { page: docs },
    jest: ['RadioButton']
  }
};

// eslint-disable-next-line react/prop-types
const RadioButtonWithState = ({ checked: initial, children, ...props }) => {
  const [checked, setChecked] = useState(initial);
  const handleChange = () => {
    setChecked(prev => !prev);
  };
  return (
    <RadioButton {...props} checked={checked} onChange={handleChange}>
      {children || (checked ? 'Checked' : 'Unchecked')}
    </RadioButton>
  );
};

export const base = () => <RadioButtonWithState value="radio" name="radio" />;

export const invalid = () => (
  <RadioButtonWithState value="invalid" name="invalid" invalid>
    Error
  </RadioButtonWithState>
);

export const disabled = () => (
  <RadioButtonWithState value="disabled" name="disabled" disabled>
    Disabled
  </RadioButtonWithState>
);
