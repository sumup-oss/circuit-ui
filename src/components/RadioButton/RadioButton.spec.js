/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import RadioButton from '.';

describe('RadioButton', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<RadioButton />);
    expect(actual).toMatchSnapshot();
  });

  it('should render with checked styles', () => {
    const actual = create(<RadioButton checked />);
    expect(actual).toMatchSnapshot();
  });

  it('should render with disabled styles', () => {
    const actual = create(<RadioButton disabled />);
    expect(actual).toMatchSnapshot();
  });

  it('should render with invalid styles', () => {
    const actual = create(<RadioButton invalid />);
    expect(actual).toMatchSnapshot();
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <RadioButton name="radio">Radio button</RadioButton>
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });

  /**
   * Logic tests.
   */
  it('should be unchecked by default', () => {
    const { getByLabelText } = render(<RadioButton>Label</RadioButton>);
    const inputEl = getByLabelText('Label', {
      exact: false
    });
    expect(inputEl).not.toHaveAttribute('checked');
  });

  it('should call the change handler when clicked', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <RadioButton onChange={onChange}>Label</RadioButton>
    );
    const inputEl = getByLabelText('Label', {
      exact: false
    });

    act(() => {
      fireEvent.click(inputEl);
    });

    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
