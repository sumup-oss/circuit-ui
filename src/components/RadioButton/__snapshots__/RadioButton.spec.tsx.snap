// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should render with checked styles 1`] = `
HTMLCollection [
  .circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-0:hover + label::before {
  border-color: #5C656F;
}

.circuit-0:focus + label::before {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-0:focus + label::before::-moz-focus-inner {
  border: 0;
}

.circuit-0:checked + label::before {
  border-color: #3388FF;
}

.circuit-0:checked + label::after {
  -webkit-transform: translateY(-50%) scale(1,1);
  -ms-transform: translateY(-50%) scale(1,1);
  transform: translateY(-50%) scale(1,1);
  opacity: 1;
}

<input
    checked=""
    class="circuit-0"
    id="radio-button_2"
    type="radio"
    value=""
  />,
  .circuit-0 {
  color: #5C656F;
  padding-left: 26px;
  position: relative;
  cursor: pointer;
}

.circuit-0::before {
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  box-shadow: 0;
  background-color: #FFFFFF;
  border: 1px solid #9DA7B1;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: border 120ms ease-in-out;
  transition: border 120ms ease-in-out;
}

.circuit-0::after {
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  background-color: #3388FF;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 4px;
  -webkit-transform: translateY(-50%) scale(0,0);
  -ms-transform: translateY(-50%) scale(0,0);
  transform: translateY(-50%) scale(0,0);
  opacity: 0;
  -webkit-transition: -webkit-transform 120ms ease-in-out, opacity 120ms ease-in-out;
  -webkit-transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
  transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
}

<label
    class="circuit-0"
    for="radio-button_2"
  />,
]
`;

exports[`RadioButton should render with default styles 1`] = `
HTMLCollection [
  .circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-0:hover + label::before {
  border-color: #5C656F;
}

.circuit-0:focus + label::before {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-0:focus + label::before::-moz-focus-inner {
  border: 0;
}

.circuit-0:checked + label::before {
  border-color: #3388FF;
}

.circuit-0:checked + label::after {
  -webkit-transform: translateY(-50%) scale(1,1);
  -ms-transform: translateY(-50%) scale(1,1);
  transform: translateY(-50%) scale(1,1);
  opacity: 1;
}

<input
    class="circuit-0"
    id="radio-button_1"
    type="radio"
    value=""
  />,
  .circuit-0 {
  color: #5C656F;
  padding-left: 26px;
  position: relative;
  cursor: pointer;
}

.circuit-0::before {
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  box-shadow: 0;
  background-color: #FFFFFF;
  border: 1px solid #9DA7B1;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: border 120ms ease-in-out;
  transition: border 120ms ease-in-out;
}

.circuit-0::after {
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  background-color: #3388FF;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 4px;
  -webkit-transform: translateY(-50%) scale(0,0);
  -ms-transform: translateY(-50%) scale(0,0);
  transform: translateY(-50%) scale(0,0);
  opacity: 0;
  -webkit-transition: -webkit-transform 120ms ease-in-out, opacity 120ms ease-in-out;
  -webkit-transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
  transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
}

<label
    class="circuit-0"
    for="radio-button_1"
  />,
]
`;

exports[`RadioButton should render with disabled styles 1`] = `
HTMLCollection [
  .circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-0:hover + label::before {
  border-color: #5C656F;
}

.circuit-0:focus + label::before {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-0:focus + label::before::-moz-focus-inner {
  border: 0;
}

.circuit-0:checked + label::before {
  border-color: #3388FF;
}

.circuit-0:checked + label::after {
  -webkit-transform: translateY(-50%) scale(1,1);
  -ms-transform: translateY(-50%) scale(1,1);
  transform: translateY(-50%) scale(1,1);
  opacity: 1;
}

<input
    class="circuit-0"
    disabled=""
    id="radio-button_3"
    type="radio"
    value=""
  />,
  .circuit-0 {
  color: #5C656F;
  padding-left: 26px;
  position: relative;
  cursor: pointer;
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-0::before {
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  box-shadow: 0;
  background-color: #FFFFFF;
  border: 1px solid #9DA7B1;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: border 120ms ease-in-out;
  transition: border 120ms ease-in-out;
}

.circuit-0::after {
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  background-color: #3388FF;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 4px;
  -webkit-transform: translateY(-50%) scale(0,0);
  -ms-transform: translateY(-50%) scale(0,0);
  transform: translateY(-50%) scale(0,0);
  opacity: 0;
  -webkit-transition: -webkit-transform 120ms ease-in-out, opacity 120ms ease-in-out;
  -webkit-transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
  transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
}

.circuit-0::before {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
  border-color: #5C656F;
  background-color: #EEF0F2;
}

.circuit-0::after {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
  background-color: #5C656F;
}

<label
    class="circuit-0"
    disabled=""
    for="radio-button_3"
  />,
]
`;

exports[`RadioButton should render with invalid styles 1`] = `
HTMLCollection [
  .circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-0:hover + label::before {
  border-color: #5C656F;
}

.circuit-0:focus + label::before {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-0:focus + label::before::-moz-focus-inner {
  border: 0;
}

.circuit-0:checked + label::before {
  border-color: #3388FF;
}

.circuit-0:checked + label::after {
  -webkit-transform: translateY(-50%) scale(1,1);
  -ms-transform: translateY(-50%) scale(1,1);
  transform: translateY(-50%) scale(1,1);
  opacity: 1;
}

.circuit-0:hover + label::before {
  border-color: #B22828;
}

.circuit-0:checked + label::before {
  border-color: #DB4D4D;
}

<input
    class="circuit-0"
    id="radio-button_4"
    type="radio"
    value=""
  />,
  .circuit-0 {
  color: #5C656F;
  padding-left: 26px;
  position: relative;
  cursor: pointer;
}

.circuit-0::before {
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  box-shadow: 0;
  background-color: #FFFFFF;
  border: 1px solid #9DA7B1;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: border 120ms ease-in-out;
  transition: border 120ms ease-in-out;
}

.circuit-0::after {
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  background-color: #3388FF;
  border-radius: 100%;
  content: '';
  display: block;
  position: absolute;
  top: 50%;
  left: 4px;
  -webkit-transform: translateY(-50%) scale(0,0);
  -ms-transform: translateY(-50%) scale(0,0);
  transform: translateY(-50%) scale(0,0);
  opacity: 0;
  -webkit-transition: -webkit-transform 120ms ease-in-out, opacity 120ms ease-in-out;
  -webkit-transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
  transition: transform 120ms ease-in-out, opacity 120ms ease-in-out;
}

.circuit-0:not(:focus)::before {
  border-color: #DB4D4D;
  background-color: #F4CBCB;
}

.circuit-0:not(:focus)::after {
  background-color: #DB4D4D;
}

<label
    class="circuit-0"
    for="radio-button_4"
  />,
]
`;
